ingest-app/src/cmr/ingest/api/generic_documents.clj

```diff
@@ -235,6 +248,117 @@
+    (when-not (and (common-concepts/is-draft-concept? draft-concept-type)
+                   (not= :variable-draft draft-concept-type))
+      (errors/throw-service-error
+       :bad-request
+       (format "Only non-variable draft can be published in this route. concept-id [%s] does not belong to a non-variable draft concept" concept-id)))
+
+    ;;Get info from metadata-db. 
+    (let [info (get-info-from-metadata-db request concept-id provider-id concept-type-in-draft)
+          request (:request info)
+          draft-native-id (:native-id info)
+          ;;publish the concept-type-in-draft
+          publish-result (case concept-type-in-draft
+                           :collection (collections/ingest-collection provider-id native-id request)
+                           :tool (tools/ingest-tool provider-id native-id request)
+                           :service (services/ingest-service provider-id native-id request)
+                           (create-generic-document request))]
+      (if (= 201 (:status publish-result))
+        ;;construct request to delete the draft.
+        (let [delete-request (-> request
+                                 (assoc-in [:route-params :native-id] draft-native-id)
+                                 (assoc-in [:route-params :concept-type] draft-concept-type)
+                                 (assoc-in [:params :native-id] draft-native-id)
+                                 (assoc-in [:params :concept-type] draft-concept-type))
+              delete-result (delete-generic-document delete-request)]
+          (if (= 200 (:status delete-result))
+            publish-result
+            (errors/throw-service-error
+             :bad-request
+             (format "Publishing draft is successful with info [%s]. Deleting draft failed with info [%s]."
+                     (:body publish-result) (:body delete-result)))))
+        publish-result))))
+    
+(defn publish-variable-draft
+  "Publish a variable draft concept. i.e. Ingest the corresponding variable and
+  delete the variable draft."
+  [provider-id native-id request coll-concept-id coll-revision-id var-draft-id]
+  (let [{:keys [draft-concept-type provider-id concept-type-in-draft]}           
+        (extract-info-from-concept-id var-draft-id) ]
+    ;; If concept-id is not a variable draft, throw error.
+    (when-not (= :variable-draft draft-concept-type)
+      (errors/throw-service-error
+       :bad-request
+       (format "Only variable draft can be published in this route. concept-id [%s] does not belong to a variable draft concept" var-draft-id)))
+
+    ;; Get information from metadata-db. 
+    (let [info (get-info-from-metadata-db request var-draft-id provider-id concept-type-in-draft)
+          request (:request info)
+          draft-native-id (:native-id info)
+          publish-result (variables/ingest-variable provider-id native-id request coll-concept-id coll-revision-id)]
+      (if (= 201 (:status publish-result))
```

Comment:
What does the error look like if you have a failed publish is that just getting handled by the `variables/ingest-variable` response?

Reply:
Just like any ingest failures.  It could be through variable/ingest-variable, tools/ingest-tool, collections/ingest-collection etc....